google-cloud-firestore/src/main/java/com/google/cloud/firestore/CollectionGroup.java

```diff
@@ -101,4 +109,60 @@ public void getPartitions(
 
     observer.onCompleted();
   }
+
+  public ApiFuture<List<QueryPartition>> getPartitions(long desiredPartitionCount) {
+    Preconditions.checkArgument(
+        desiredPartitionCount > 0, "Desired partition count must be one or greater");
+
+    // Partition queries require explicit ordering by __name__.
+    final Query queryWithDefaultOrder = orderBy(FieldPath.DOCUMENT_ID);
+
+    if (desiredPartitionCount == 1) {
+      // Short circuit if the user only requested a single partition.
+      return ApiFutures.immediateFuture(
+          Collections.singletonList(new QueryPartition(queryWithDefaultOrder, null, null)));
+    } else {
+      PartitionQueryRequest.Builder request = PartitionQueryRequest.newBuilder();
+      request.setStructuredQuery(queryWithDefaultOrder.buildQuery());
+      request.setParent(options.getParentPath().toString());
+
+      // Since we are always returning an extra partition (with en empty endBefore cursor), we
+      // reduce the desired partition count by one.
+      request.setPartitionCount(desiredPartitionCount - 1);
+
+      final TraceUtil traceUtil = TraceUtil.getInstance();
```

Comment:
All the lines above here could be factored out into a helper and shared.

Reply:
I could pull out up to the PartitionQueryRequest.Builder line, but further than that and I get into the territory have having to model an abstract lifecycle type object to allow the individual handling of the single case. Which also would extend to the loop below. If we had some existing "addable" or similar type we could use it to wrap the observer and the list but I don't think it's worth it here.

What do you think, should I still work to factor this out?

Reply:
I played a bit with this (but didn't run a test): https://gist.github.com/schmidt-sebastian/ce122b16bd13837cf9c01cfe38783bf1

Let me know what you think.

Reply:
I see, thanks for sharing the patch I've incorporated it into the PR.